File: java/tags/direct-policy-1.0/src/test/java/org/nhindirect/policy/impl/BinaryBooleanPolicyOperatorExecutor_executeTest.java
package org.nhindirect.policy.impl;

import java.util.Arrays;
import java.util.List;

import org.nhindirect.policy.PolicyOperator;
import org.nhindirect.policy.PolicyValue;
import org.nhindirect.policy.PolicyValueFactory;

import junit.framework.TestCase;

public class BinaryBooleanPolicyOperatorExecutor_executeTest extends TestCase
{
	public void testExecute_logicalAnd_assertResults()
	{
		// both true
		PolicyValue<Boolean> op1 = PolicyValueFactory.getInstance(true);
		PolicyValue<Boolean> op2 = PolicyValueFactory.getInstance(true);
		
		BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean> operator = new BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean>(op1, op2, PolicyOperator.LOGICAL_AND);
		assertTrue(operator.execute().getPolicyValue());
		
		// one false one true
		op1 = PolicyValueFactory.getInstance(true);
		op2 = PolicyValueFactory.getInstance(false);
		
		operator = new BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean>(op1, op2, PolicyOperator.LOGICAL_AND);
		assertFalse(operator.execute().getPolicyValue());
	
		// both false
		op1 = PolicyValueFactory.getInstance(false);
		op2 = PolicyValueFactory.getInstance(false);
		
		operator = new BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean>(op1, op2, PolicyOperator.LOGICAL_AND);
		assertFalse(operator.execute().getPolicyValue());
	}
	
	public void testExecute_logicalAnd_stringArguments_assertResults()
	{
		// both true
		PolicyValue<String> op1 = PolicyValueFactory.getInstance("true");
		PolicyValue<String> op2 = PolicyValueFactory.getInstance("true");
		
		BinaryBooleanPolicyOperatorExecutor<String, String> operator = new BinaryBooleanPolicyOperatorExecutor<String, String>(op1, op2, PolicyOperator.LOGICAL_AND);
		assertTrue(operator.execute().getPolicyValue());
		
		// one false one true
		op1 = PolicyValueFactory.getInstance("true");
		op2 = PolicyValueFactory.getInstance("false");
		
		operator = new BinaryBooleanPolicyOperatorExecutor<String, String>(op1, op2, PolicyOperator.LOGICAL_AND);
		assertFalse(operator.execute().getPolicyValue());
	
		// both false
		op1 = PolicyValueFactory.getInstance("false");
		op2 = PolicyValueFactory.getInstance("false");
		
		operator = new BinaryBooleanPolicyOperatorExecutor<String, String>(op1, op2, PolicyOperator.LOGICAL_AND);
		assertFalse(operator.execute().getPolicyValue());
	}
	
	public void testExecute_logicalOr_assertResults()
	{
		// both true
		PolicyValue<Boolean> op1 = PolicyValueFactory.getInstance(true);
		PolicyValue<Boolean> op2 = PolicyValueFactory.getInstance(true);
		
		BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean> operator = new BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean>(op1, op2, PolicyOperator.LOGICAL_OR);
		assertTrue(operator.execute().getPolicyValue());
		
		// one false one true
		op1 = PolicyValueFactory.getInstance(true);
		op2 = PolicyValueFactory.getInstance(false);
		
		operator = new BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean>(op1, op2, PolicyOperator.LOGICAL_OR);
		assertTrue(operator.execute().getPolicyValue());
	
		// both false
		op1 = PolicyValueFactory.getInstance(false);
		op2 = PolicyValueFactory.getInstance(false);
		
		operator = new BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean>(op1, op2, PolicyOperator.LOGICAL_OR);
		assertFalse(operator.execute().getPolicyValue());
	}	
	
	public void testExecute_equals_assertResults()
	{
		// boolean equals
		PolicyValue<Boolean> op1 = PolicyValueFactory.getInstance(true);
		PolicyValue<Boolean> op2 = PolicyValueFactory.getInstance(true);
		
		BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean> operator = new BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean>(op1, op2, PolicyOperator.EQUALS);
		assertTrue(operator.execute().getPolicyValue());
		
		// boolean not equals
		op1 = PolicyValueFactory.getInstance(true);
		op2 = PolicyValueFactory.getInstance(false);
		
		operator = new BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean>(op1, op2, PolicyOperator.EQUALS);
		assertFalse(operator.execute().getPolicyValue());
	
		// integer equals
		PolicyValue<Integer> opInt1 = PolicyValueFactory.getInstance(123);
		PolicyValue<Integer> opInt2 = PolicyValueFactory.getInstance(123);
		
		BinaryBooleanPolicyOperatorExecutor<Integer, Integer> intOperator = new BinaryBooleanPolicyOperatorExecutor<Integer, Integer>(opInt1, opInt2, PolicyOperator.EQUALS);
		assertTrue(intOperator.execute().getPolicyValue());
		
		// integer not equals
		opInt1 = PolicyValueFactory.getInstance(123);
		opInt2 = PolicyValueFactory.getInstance(456);
		
		intOperator = new BinaryBooleanPolicyOperatorExecutor<Integer, Integer>(opInt1, opInt2, PolicyOperator.EQUALS);
		assertFalse(intOperator.execute().getPolicyValue());
	}	
	
	public void testExecute_notEquals_assertResults()
	{
		// boolean equals
		PolicyValue<Boolean> op1 = PolicyValueFactory.getInstance(true);
		PolicyValue<Boolean> op2 = PolicyValueFactory.getInstance(true);
		
		BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean> operator = new BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean>(op1, op2, PolicyOperator.NOT_EQUALS);
		assertFalse(operator.execute().getPolicyValue());
		
		// boolean not equals
		op1 = PolicyValueFactory.getInstance(true);
		op2 = PolicyValueFactory.getInstance(false);
		
		operator = new BinaryBooleanPolicyOperatorExecutor<Boolean, Boolean>(op1, op2, PolicyOperator.NOT_EQUALS);
		assertTrue(operator.execute().getPolicyValue());
	
		// integer equals
		PolicyValue<Integer> opInt1 = PolicyValueFactory.getInstance(123);
		PolicyValue<Integer> opInt2 = PolicyValueFactory.getInstance(123);
		
		BinaryBooleanPolicyOperatorExecutor<Integer, Integer> intOperator = new BinaryBooleanPolicyOperatorExecutor<Integer, Integer>(opInt1, opInt2, PolicyOperator.NOT_EQUALS);
		assertFalse(intOperator.execute().getPolicyValue());
		
		// integer not equals
		opInt1 = PolicyValueFactory.getInstance(123);
		opInt2 = PolicyValueFactory.getInstance(456);
		
		intOperator = new BinaryBooleanPolicyOperatorExecutor<Integer, Integer>(opInt1, opInt2, PolicyOperator.NOT_EQUALS);
		assertTrue(intOperator.execute().getPolicyValue());
	}	
	
	public void testExecute_greater_assertResults()
	{	
		// op1 greater than op2
		// compiler uses reverse polish notation, so the operands need to be backwards
		PolicyValue<Integer> opInt1 = PolicyValueFactory.getInstance(5);
		PolicyValue<Integer> opInt2 = PolicyValueFactory.getInstance(4);
		
		BinaryBooleanPolicyOperatorExecutor<Integer, Integer> intOperator = new BinaryBooleanPolicyOperatorExecutor<Integer, Integer>(opInt1, opInt2, PolicyOperator.GREATER);
		assertFalse(intOperator.execute().getPolicyValue());
		
		// op1 less than op2
		opInt1 = PolicyValueFactory.getInstance(4);
		opInt2 = PolicyValueFactory.getInstance(5);
		
		intOperator = new BinaryBooleanPolicyOperatorExecutor<Integer, Integer>(opInt1, opInt2, PolicyOperator.GREATER);
		assertTrue(intOperator.execute().getPolicyValue());
	}		
	
	public void testExecute_less_assertResults()
	{	
		// op1 less than op2
		// compiler uses reverse polish notation, so the operands need to be backwards
		PolicyValue<Integer> opInt1 = PolicyValueFactory.getInstance(4);
		PolicyValue<Integer> opInt2 = PolicyValueFactory.getInstance(5);
		
		BinaryBooleanPolicyOperatorExecutor<Integer, Integer> intOperator = new BinaryBooleanPolicyOperatorExecutor<Integer, Integer>(opInt1, opInt2, PolicyOperator.LESS);
		assertFalse(intOperator.execute().getPolicyValue());
		
		// op1 greater than op2
		opInt1 = PolicyValueFactory.getInstance(5);
		opInt2 = PolicyValueFactory.getInstance(4);
		
		intOperator = new BinaryBooleanPolicyOperatorExecutor<Integer, Integer>(opInt1, opInt2, PolicyOperator.LESS);
		assertTrue(intOperator.execute().getPolicyValue());
	}	
	
	public void testExecute_contains_assertResults()
	{	
		// container contains string
		PolicyValue<String> op1 = PolicyValueFactory.getInstance("689");
		PolicyValue<List<String>> op2 = PolicyValueFactory.getInstance(Arrays.asList("123","456","689"));
		
		BinaryBooleanPolicyOperatorExecutor<String, List<String>> operator = 
				new BinaryBooleanPolicyOperatorExecutor<String, List<String>>(op1, op2, PolicyOperator.CONTAINS);
		assertTrue(operator.execute().getPolicyValue());
		
		// container does not contain string
		op1 = PolicyValueFactory.getInstance("777");
		op2 = PolicyValueFactory.getInstance(Arrays.asList("123","456","689"));
		
		operator = new BinaryBooleanPolicyOperatorExecutor<String, List<String>>(op1, op2, PolicyOperator.CONTAINS);
		assertFalse(operator.execute().getPolicyValue());
	}	
	
	public void testExecute_notContains_assertResults()
	{	
		// container contains string
		PolicyValue<String> op1 = PolicyValueFactory.getInstance("689");
		PolicyValue<List<String>> op2 = PolicyValueFactory.getInstance(Arrays.asList("123","456","689"));
		
		BinaryBooleanPolicyOperatorExecutor<String, List<String>> operator = 
				new BinaryBooleanPolicyOperatorExecutor<String, List<String>>(op1, op2, PolicyOperator.NOT_CONTAINS);
		assertFalse(operator.execute().getPolicyValue());
		
		// container does not contain string
		op1 = PolicyValueFactory.getInstance("777");
		op2 = PolicyValueFactory.getInstance(Arrays.asList("123","456","689"));
		
		operator = new BinaryBooleanPolicyOperatorExecutor<String, List<String>>(op1, op2, PolicyOperator.NOT_CONTAINS);
		assertTrue(operator.execute().getPolicyValue());
	}		
	
	public void testExecute_containsRegEx_assertResults()
	{	
		// container contains string reg ex
		PolicyValue<String> op1 = PolicyValueFactory.getInstance("http");
		PolicyValue<List<String>> op2 = PolicyValueFactory.getInstance(Arrays.asList("http://thisis.aurl.com"));
		
		BinaryBooleanPolicyOperatorExecutor<String, List<String>> operator = 
				new BinaryBooleanPolicyOperatorExecutor<String, List<String>>(op1, op2, PolicyOperator.CONTAINS_REG_EX);
		assertTrue(operator.execute().getPolicyValue());
		
		// container does not contain string reg ex
		op1 = PolicyValueFactory.getInstance("777");
		op2 = PolicyValueFactory.getInstance(Arrays.asList("http://thisis.aurl.com"));
		
		operator = new BinaryBooleanPolicyOperatorExecutor<String, List<String>>(op1, op2, PolicyOperator.CONTAINS_REG_EX);
		assertFalse(operator.execute().getPolicyValue());
	}	
	
	public void testExecute_regEx_assertResults()
	{	
		// string contains string reg ex
		PolicyValue<String> op1 = PolicyValueFactory.getInstance("http");
		PolicyValue<String> op2 = PolicyValueFactory.getInstance("http://thisis.aurl.com");
		
		BinaryBooleanPolicyOperatorExecutor<String, String> operator = 
				new BinaryBooleanPolicyOperatorExecutor<String, String>(op1, op2, PolicyOperator.REG_EX);
		assertTrue(operator.execute().getPolicyValue());
		
		// container does not contain string reg ex
		op1 = PolicyValueFactory.getInstance("777");
		op2 = PolicyValueFactory.getInstance("http://thisis.aurl.com");
		
		operator = new BinaryBooleanPolicyOperatorExecutor<String, String>(op1, op2, PolicyOperator.REG_EX);
		assertFalse(operator.execute().getPolicyValue());
	}		
}

File: csharp/unittests/policy.tests/BinaryBooleanPolicyOperatorExecutor_executeTest.cs
/* 
 Copyright (c) 2013, Direct Project
 All rights reserved.

 Authors:
    Joe Shook      jshook@kryptiq.com
  
Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
Neither the name of The Direct Project (directproject.org) nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
*/


using System;
using System.Collections.Generic;
using System.Linq;
using FluentAssertions;
using Health.Direct.Policy.Extensions;
using Health.Direct.Policy.Operators;
using Xunit;

namespace Health.Direct.Policy.Tests
{
    public class BinaryBooleanPolicyOperatorExecutor_ExecuteTest
    {
        [Fact]
        public void TestExecute_LogicalAnd_AssertResults()
        {
            Assert.True(PolicyOperator<bool>.LOGICAL_AND.Execute(true, true));
            Assert.False(PolicyOperator<bool>.LOGICAL_AND.Execute(true, false));
            Assert.False(PolicyOperator<bool>.LOGICAL_AND.Execute(false, true));
            Assert.False(PolicyOperator<bool>.LOGICAL_AND.Execute(false, false)); 
        }

        [Fact]
        public void TestExecute_LogicalAnd_StringArguments_AssertResults()
        {
            //don't care... maybe...
        }

        [Fact]
        public void TestExecute_LogicalOr_AssertResults()
        {
            OperatorBase equalsOperator = PolicyOperator.FromToken(("=_" + "Boolean_Boolean").GetHashCode());
            equalsOperator.Should().NotBeNull();

            Assert.True(PolicyOperator<bool>.LOGICAL_OR.Execute(true, true));
            Assert.True(PolicyOperator<bool>.LOGICAL_OR.Execute(true, false));
            Assert.True(PolicyOperator<bool>.LOGICAL_OR.Execute(false, true));
            Assert.False(PolicyOperator<bool>.LOGICAL_OR.Execute(false, false)); 
        }

        [Fact]
        public void TestExecute_Equals_AssertResults()
        {
            Assert.True(PolicyOperator<bool, bool, bool>.EQUALS.Execute(true, true));
            Assert.False(PolicyOperator<bool, bool, bool>.EQUALS.Execute(true, false));
            Assert.True(PolicyOperator<int, int, bool>.EQUALS.Execute(123, 123));
            Assert.False(PolicyOperator<int, int, bool>.EQUALS.Execute(123, 456)); 
        }

        [Fact]
        public void TestExecute_Equals_Convertparamspecial_AssertResults()
        {
            Assert.True(PolicyOperator<Int64, Int64, bool>.EQUALS.Execute("00F74F1C4FE4E1762E".HexAsLong(), "f74f1c4fe4e1762e".HexAsLong()));
            Assert.True(PolicyOperator<Int64, String, bool>.EQUALS.Execute("00F74F1C4FE4E1762E".HexAsLong(), "f74f1c4fe4e1762e"));

            Delegate del = PolicyOperator<Int64, String, bool>.EQUALS.ExecuteRef;
            Assert.True((bool)del.DynamicInvoke(new object[] { "00F74F1C4FE4E1762E".HexAsLong(), "f74f1c4fe4e1762e" }));

        }

        [Fact]
        public void TestExecute_Equals_Convertparam_AssertResults()
        {
            Assert.True(PolicyOperator<int, int, bool>.EQUALS.Execute(123, 123));
            Assert.True(PolicyOperator<int, String, bool>.EQUALS.Execute(123, "123"));

            Delegate del = PolicyOperator<int, String, bool>.EQUALS.ExecuteRef;
            Assert.True((bool)del.DynamicInvoke(new object[] { 123, "123"}));

        }


        [Fact]
        public void TestExecute_Equals_Dynamic_AssertResults()
        {
            Delegate del = PolicyOperator<bool, bool, bool>.EQUALS.ExecuteRef;

            Assert.True((bool)del.DynamicInvoke(new object[] { true, true }));
            Assert.False((bool)del.DynamicInvoke(new object[] { true, false }));

            del = PolicyOperator<int, int, bool>.EQUALS.ExecuteRef;
            Assert.True((bool)del.DynamicInvoke(new object[] { 123, 123 }));
            Assert.False((bool)del.DynamicInvoke(new object[] { 123, 456 }));
        }

        [Fact]
        public void TestExecute_NotEquals_AssertResults()
        {
            Assert.False(PolicyOperator<bool, bool, bool>.NOT_EQUALS.Execute(true, true));
            Assert.True(PolicyOperator<bool, bool, bool>.NOT_EQUALS.Execute(true, false));
            Assert.False(PolicyOperator<int, int, bool>.NOT_EQUALS.Execute(123, 123));
            Assert.True(PolicyOperator<int, int, bool>.NOT_EQUALS.Execute(123, 456)); 
        }

        [Fact]
        public void TestExecute_NotEquals_Convertparamspecial_AssertResults()
        {
            Assert.False(PolicyOperator<Int64, Int64, bool>.NOT_EQUALS.Execute("00F74F1C4FE4E1762E".HexAsLong(), "f74f1c4fe4e1762e".HexAsLong()));
            Assert.False(PolicyOperator<Int64, String, bool>.NOT_EQUALS.Execute("00F74F1C4FE4E1762E".HexAsLong(), "f74f1c4fe4e1762e"));
            Assert.True(PolicyOperator<Int64, String, bool>.NOT_EQUALS.Execute("00F74F1C4FE4E17600".HexAsLong(), "f74f1c4fe4e1762e"));

            Delegate del = PolicyOperator<Int64, String, bool>.NOT_EQUALS.ExecuteRef;
            Assert.False((bool) del.DynamicInvoke(new object[] {"00F74F1C4FE4E1762E".HexAsLong(), "f74f1c4fe4e1762e"}));
            Assert.True((bool)del.DynamicInvoke(new object[] { "00F74F1C4FE4E17600".HexAsLong(), "f74f1c4fe4e1762e" }));

        }

        [Fact]
        public void TestExecute_NotEquals_Convertparam_AssertResults()
        {
            Assert.False(PolicyOperator<int, int, bool>.NOT_EQUALS.Execute(123, 123));
            Assert.False(PolicyOperator<int, String, bool>.NOT_EQUALS.Execute(123, "123"));

            Delegate del = PolicyOperator<int, String, bool>.NOT_EQUALS.ExecuteRef;
            Assert.False((bool)del.DynamicInvoke(new object[] { 123, "123" }));

        }


        [Fact]
        public void TestExecute_NotEquals_Dynamic_AssertResults()
        {
            Delegate del = PolicyOperator<bool, bool, bool>.NOT_EQUALS.ExecuteRef;

            Assert.False((bool)del.DynamicInvoke(new object[] { true, true }));
            Assert.True((bool)del.DynamicInvoke(new object[] { true, false }));

            del = PolicyOperator<int, int, bool>.NOT_EQUALS.ExecuteRef;
            Assert.False((bool)del.DynamicInvoke(new object[] { 123, 123 }));
            Assert.True((bool)del.DynamicInvoke(new object[] { 123, 456 }));
        }


        [Fact]
        public void TestExecute_Greater_AssertResults()
        {
            Assert.True(PolicyOperator<int, int>.GREATER.Execute(5, 4));
            Assert.False(PolicyOperator<int, int>.GREATER.Execute(4, 5));
        }

        [Fact]
        public void TestExecute_Less_AssertResults()
        {
            Assert.True(PolicyOperator<int, int>.LESS.Execute(4, 5));
            Assert.False(PolicyOperator<int, int>.LESS.Execute(5, 4));
        }

        [Fact]
        public void TestExecute_Contains_AssertResults()
        {
            var strings = new List<string> {"123", "456", "689"};
            strings.Any(m => m == "123").Should().BeTrue();
            Assert.True(PolicyOperator<IList<string>, String, bool>.CONTAINS.Execute(strings, "689"));
            Assert.True(PolicyOperator<IList<string>, String, bool>.CONTAINS.Execute(strings, "456"));
            Assert.False(PolicyOperator<IList<string>, String, bool>.CONTAINS.Execute(strings, "777"));

            var integers = new List<int> { 123, 456, 689 };
            Assert.True(PolicyOperator<IList<int>, int, bool>.CONTAINS.Execute(integers, 689));
            Assert.True(PolicyOperator<IList<int>, int, bool>.CONTAINS.Execute(integers, 456));
            Assert.False(PolicyOperator<IList<int>, int, bool>.CONTAINS.Execute(integers, 777));
        }

        

        [Fact]
        public void TestExecute_NotContains_AssertResults()
        {
            var strings = new List<string> { "123", "456", "689" };
            Assert.False(PolicyOperator<IList<string>, String, bool>.NOT_CONTAINS.Execute(strings, "689"));
            Assert.False(PolicyOperator<IList<string>, String, bool>.NOT_CONTAINS.Execute(strings, "456"));
            Assert.True(PolicyOperator<IList<string>, String, bool>.NOT_CONTAINS.Execute(strings, "777"));

            var integers  = new List<int> { 123, 456, 689 };
            Assert.False(PolicyOperator<IList<int>, int, bool>.NOT_CONTAINS.Execute(integers, 689));
            Assert.False(PolicyOperator<IList<int>, int, bool>.NOT_CONTAINS.Execute(integers, 456));
            Assert.True(PolicyOperator<IList<int>, int, bool>.NOT_CONTAINS.Execute(integers, 777));
        }


        [Fact]
        public void TestExecute_ContainsRegEx_AssertResults()
        {
            var urls = new List<string> {"http://thisis.aurl.com"};
            Assert.True(PolicyOperator<IList<String>, String, bool>.CONTAINS_REG_EX.Execute(urls, "http"));

            Assert.False(PolicyOperator<IList<String>, String, bool>.CONTAINS_REG_EX.Execute(urls, "777"));

        }
        [Fact]
        public void TestExecute_RegEx_AssertResults()
        {
            Assert.True(PolicyOperator<string, bool>
                .REG_EX.Execute("http", "http://thisis.aurl.com"));

            Assert.False(PolicyOperator<string, bool>
                .REG_EX.Execute("777", "http://thisis.aurl.com"));
        }



    }
    
}
